File: spark/so/grpc_test/coop_exit_test.go
package grpctest

import (
	"bytes"
	"context"
	"testing"
	"time"

	"github.com/btcsuite/btcd/wire"
	"github.com/decred/dcrd/dcrec/secp256k1/v4"
	"github.com/lightsparkdev/spark-go/common"
	"github.com/lightsparkdev/spark-go/proto/spark"
	"github.com/lightsparkdev/spark-go/so/handler"
	testutil "github.com/lightsparkdev/spark-go/test_util"
	"github.com/lightsparkdev/spark-go/wallet"
	"github.com/stretchr/testify/assert"
)

func TestCoopExit(t *testing.T) {
	client, err := testutil.NewRegtestClient()
	assert.NoError(t, err)

	coin, err := faucet.Fund()
	assert.NoError(t, err)

	config, err := testutil.TestWalletConfig()
	assert.NoError(t, err)

	amountSats := int64(100_000) // TODO: this should match the amount from the leaves

	// Setup a user with some leaves
	leafPrivKey, err := secp256k1.GeneratePrivateKey()
	assert.NoError(t, err)
	rootNode, err := testutil.CreateNewTree(config, faucet, leafPrivKey, amountSats)
	assert.NoError(t, err)

	time.Sleep(100 * time.Millisecond)

	// Initiate SSP
	sspPrivKey, err := secp256k1.GeneratePrivateKey()
	assert.NoError(t, err)
	sspPubkey := sspPrivKey.PubKey()
	sspIntermediateAddress, err := common.P2TRAddressFromPublicKey(sspPubkey.SerializeCompressed(), config.Network)
	assert.NoError(t, err)
	sspConfig, err := testutil.TestWalletConfigWithIdentityKey(*sspPrivKey)
	assert.NoError(t, err)

	// Initiate exit - SSP is just another user, providing a service external to the SO
	withdrawPrivKey, err := secp256k1.GeneratePrivateKey()
	assert.NoError(t, err)
	withdrawPubKey := withdrawPrivKey.PubKey()
	withdrawAddress, err := common.P2TRAddressFromPublicKey(withdrawPubKey.SerializeCompressed(), config.Network)
	assert.NoError(t, err)

	leafCount := 1                                                    // TODO: this should match the number of leaves
	dustAmountSats := 354                                             // TODO: this should match the proper dust
	intermediateAmountSats := int64((leafCount + 1) * dustAmountSats) // +1 for an output SSP can fee bump

	exitTx, err := testutil.CreateTestCoopExitTransaction(coin.OutPoint, *withdrawAddress, amountSats, *sspIntermediateAddress, intermediateAmountSats)
	assert.NoError(t, err)

	exitTxHash := exitTx.TxHash()
	intermediateOutPoint := wire.NewOutPoint(&exitTxHash, 1)
	connectorP2trAddrs := make([]string, 0)
	for i := 0; i < leafCount+1; i++ {
		connectorPrivKey, err := secp256k1.GeneratePrivateKey()
		assert.NoError(t, err)
		connectorPubKey := connectorPrivKey.PubKey()
		connectorAddress, err := common.P2TRAddressFromPublicKey(connectorPubKey.SerializeCompressed(), config.Network)
		assert.NoError(t, err)
		connectorP2trAddrs = append(connectorP2trAddrs, *connectorAddress)
	}
	feeBumpAddr := connectorP2trAddrs[len(connectorP2trAddrs)-1]
	connectorP2trAddrs = connectorP2trAddrs[:len(connectorP2trAddrs)-1]
	connectorTx, err := testutil.CreateTestConnectorTransaction(intermediateOutPoint, intermediateAmountSats, connectorP2trAddrs, feeBumpAddr)
	assert.NoError(t, err)

	connectorOutputs := make([]*wire.OutPoint, 0)
	for i := range connectorTx.TxOut[:len(connectorTx.TxOut)-1] {
		txHash := connectorTx.TxHash()
		connectorOutputs = append(connectorOutputs, wire.NewOutPoint(&txHash, uint32(i)))
	}

	newLeafPrivKey, err := secp256k1.GeneratePrivateKey()
	assert.NoError(t, err)

	transferNode := wallet.LeafKeyTweak{
		Leaf:              rootNode,
		SigningPrivKey:    leafPrivKey.Serialize(),
		NewSigningPrivKey: newLeafPrivKey.Serialize(),
	}

	senderTransfer, _, err := wallet.GetConnectorRefundSignatures(
		context.Background(),
		config,
		[]wallet.LeafKeyTweak{transferNode},
		exitTxHash.CloneBytes(),
		connectorOutputs,
		sspPubkey,
	)
	assert.NoError(t, err)
	assert.Equal(t, senderTransfer.Status, spark.TransferStatus_TRANSFER_STATUS_SENDER_KEY_TWEAK_PENDING)

	sspToken, err := wallet.AuthenticateWithServer(context.Background(), sspConfig)
	assert.NoError(t, err)
	sspCtx := wallet.ContextWithToken(context.Background(), sspToken)

	// Try to claim leaf before exit tx confirms -> should fail
	finalLeafPrivKey, err := secp256k1.GeneratePrivateKey()
	assert.NoError(t, err)
	claimingNode := wallet.LeafKeyTweak{
		Leaf:              senderTransfer.Leaves[0].Leaf,
		SigningPrivKey:    newLeafPrivKey.Serialize(),
		NewSigningPrivKey: finalLeafPrivKey.Serialize(),
	}
	leavesToClaim := [1]wallet.LeafKeyTweak{claimingNode}
	_, err = wallet.ClaimTransfer(
		sspCtx,
		senderTransfer,
		sspConfig,
		leavesToClaim[:],
	)
	if err == nil {
		t.Fatalf("expected error claiming transfer before exit tx confirms")
	}

	// Sign exit tx and broadcast
	signedExitTx, err := testutil.SignFaucetCoin(exitTx, coin.TxOut, coin.Key)
	assert.NoError(t, err)

	_, err = client.SendRawTransaction(signedExitTx, true)
	assert.NoError(t, err)

	// Make sure the exit tx gets enough confirmations
	randomKey, err := secp256k1.GeneratePrivateKey()
	assert.NoError(t, err)
	randomPubKey := randomKey.PubKey()
	randomAddress, err := common.P2TRRawAddressFromPublicKey(randomPubKey.SerializeCompressed(), common.Regtest)
	assert.NoError(t, err)
	// Confirm extra buffer to scan more blocks than needed
	// So that we don't race the chain watcher in this test
	_, err = client.GenerateToAddress(handler.CoopExitConfirmationThreshold+6, randomAddress, nil)
	assert.NoError(t, err)

	// Wait until tx is confirmed and picked up by SO
	pendingTransfer, err := wallet.QueryPendingTransfers(sspCtx, sspConfig)
	assert.NoError(t, err)
	startTime := time.Now()
	for len(pendingTransfer.Transfers) == 0 {
		if time.Since(startTime) > 10*time.Second {
			t.Fatalf("timed out waiting for key to be tweaked from tx confirmation")
		}
		time.Sleep(100 * time.Millisecond)
		pendingTransfer, err = wallet.QueryPendingTransfers(sspCtx, sspConfig)
		assert.NoError(t, err)
	}
	receiverTransfer := pendingTransfer.Transfers[0]
	assert.Equal(t, receiverTransfer.Id, senderTransfer.Id)
	assert.Equal(t, receiverTransfer.Status, spark.TransferStatus_TRANSFER_STATUS_SENDER_KEY_TWEAKED)

	leafPrivKeyMap, err := wallet.VerifyPendingTransfer(context.Background(), sspConfig, receiverTransfer)
	assert.NoError(t, err)
	assert.Equal(t, len(*leafPrivKeyMap), 1)
	assert.True(t, bytes.Equal((*leafPrivKeyMap)[rootNode.Id], newLeafPrivKey.Serialize()))

	// Claim leaf. This requires a loop because sometimes there are
	// delays in processing blocks, and after the tx initially confirms,
	// the SO will still reject a claim until the tx has enough confirmations.
	startTime = time.Now()
	for {
		_, err = wallet.ClaimTransfer(
			sspCtx,
			receiverTransfer,
			sspConfig,
			leavesToClaim[:],
		)
		if err == nil {
			break
		}
		time.Sleep(200 * time.Millisecond)
		if time.Since(startTime) > 15*time.Second {
			t.Fatalf("timed out waiting for tx to confirm")
		}
	}
}

File: sdks/spark-js-sdk/src/tests/coop-exit.test.ts
import { describe, expect, it } from "@jest/globals";
import { hexToBytes } from "@noble/curves/abstract/utils";
import { secp256k1 } from "@noble/curves/secp256k1";
import { Address, OutScript, Transaction } from "@scure/btc-signer";
import { TransactionInput } from "@scure/btc-signer/psbt";
import { equalBytes, sha256 } from "@scure/btc-signer/utils";
import { WalletConfigService } from "../services/config";
import { ConnectionManager } from "../services/connection";
import { CoopExitService } from "../services/coop-exit";
import { LeafKeyTweak } from "../services/transfer";
import { SparkWallet } from "../spark-sdk";
import {
  getP2TRAddressFromPublicKey,
  getP2TRScriptFromPublicKey,
  getTxId,
  getTxIdNoReverse,
} from "../utils/bitcoin";
import { getNetwork, Network } from "../utils/network";
import { createNewTree } from "./test-util";
import { BitcoinFaucet } from "./utils/test-faucet";

describe("coop exit", () => {
  // Skip all tests if running in GitHub Actions
  const testFn = process.env.GITHUB_ACTIONS ? it.skip : it;

  testFn(
    "test coop exit",
    async () => {
      const faucet = new BitcoinFaucet(
        "http://127.0.0.1:18443",
        "admin1",
        "123"
      );

      const faucetCoin = await faucet.fund();

      const amountSats = 100_000n;

      // Setup user with leaves
      const userWallet = new SparkWallet(Network.REGTEST);
      const userMnemonic = userWallet.generateMnemonic();
      await userWallet.createSparkWallet(userMnemonic);

      const configService = new WalletConfigService(
        Network.REGTEST,
        userWallet.getSigner()
      );
      const connectionManager = new ConnectionManager(configService);
      const coopExitService = new CoopExitService(
        configService,
        connectionManager
      );

      const leafPubKey = userWallet
        .getSigner()
        .generatePublicKey(sha256("leafPubKey"));
      const rootNode = await createNewTree(
        userWallet,
        leafPubKey,
        faucet,
        amountSats
      );

      // Setup ssp
      const sspWallet = new SparkWallet(Network.REGTEST);
      const sspMnemonic = sspWallet.generateMnemonic();
      const sspPubkey = await sspWallet.createSparkWallet(sspMnemonic);

      const sspIntermediateAddressScript = getP2TRScriptFromPublicKey(
        hexToBytes(sspPubkey),
        Network.REGTEST
      );

      // Setup withdraw
      const withdrawPubKey = userWallet.getSigner().generatePublicKey();
      const withdrawAddressScript = getP2TRScriptFromPublicKey(
        withdrawPubKey,
        Network.REGTEST
      );

      const leafCount = 1;
      const dustAmountSats = 354;
      const intermediateAmountSats = (leafCount + 1) * dustAmountSats;

      const exitTx = new Transaction();
      exitTx.addInput(faucetCoin.outpoint);
      exitTx.addOutput({
        script: withdrawAddressScript,
        amount: amountSats,
      });
      exitTx.addOutput({
        script: sspIntermediateAddressScript,
        amount: BigInt(intermediateAmountSats),
      });

      const exitTxId = getTxId(exitTx);
      const intermediateOutPoint: TransactionInput = {
        txid: hexToBytes(exitTxId),
        index: 1,
      };

      let connectorP2trAddrs: string[] = [];
      for (let i = 0; i < leafCount + 1; i++) {
        const connectorPubKey = userWallet.getSigner().generatePublicKey();
        const connectorP2trAddr = getP2TRAddressFromPublicKey(
          connectorPubKey,
          Network.REGTEST
        );
        connectorP2trAddrs.push(connectorP2trAddr);
      }
      const feeBumpAddr = connectorP2trAddrs[connectorP2trAddrs.length - 1];
      connectorP2trAddrs = connectorP2trAddrs.slice(0, -1);

      const connectorTx = new Transaction();
      connectorTx.addInput(intermediateOutPoint);
      for (const addr of [...connectorP2trAddrs, feeBumpAddr]) {
        connectorTx.addOutput({
          script: OutScript.encode(
            Address(getNetwork(Network.REGTEST)).decode(addr)
          ),
          amount: BigInt(intermediateAmountSats / connectorP2trAddrs.length),
        });
      }

      const connectorOutputs: TransactionInput[] = [];
      for (let i = 0; i < connectorTx.outputsLength - 1; i++) {
        connectorOutputs.push({
          txid: hexToBytes(getTxId(connectorTx)),
          index: i,
        });
      }

      const newLeafPubKey = userWallet.getSigner().generatePublicKey();

      const transferNode = {
        leaf: rootNode,
        signingPubKey: leafPubKey,
        newSigningPubKey: newLeafPubKey,
      };

      const senderTransfer = await coopExitService.getConnectorRefundSignatures(
        {
          leaves: [transferNode],
          exitTxId: hexToBytes(getTxIdNoReverse(exitTx)),
          connectorOutputs,
          receiverPubKey: hexToBytes(sspPubkey),
        }
      );

      const pendingTransfer = await sspWallet.queryPendingTransfers();

      expect(pendingTransfer.transfers.length).toBe(1);

      const receiverTransfer = pendingTransfer.transfers[0];
      expect(receiverTransfer.id).toBe(senderTransfer.transfer.id);

      const leafPubKeyMap = await sspWallet.verifyPendingTransfer(
        receiverTransfer
      );

      expect(leafPubKeyMap.size).toBe(1);
      expect(leafPubKeyMap.get(rootNode.id)).toBeDefined();
      expect(equalBytes(leafPubKeyMap.get(rootNode.id)!, newLeafPubKey)).toBe(
        true
      );

      // Try to claim leaf before exit tx confirms -> should fail
      const finalLeafPubKey = sspWallet
        .getSigner()
        .generatePublicKey(sha256("finalLeafPubKey"));

      const leavesToClaim: LeafKeyTweak[] = [
        {
          leaf: receiverTransfer.leaves[0].leaf!,
          signingPubKey: newLeafPubKey,
          newSigningPubKey: finalLeafPubKey,
        },
      ];

      let hasError = false;
      try {
        await sspWallet._claimTransfer(receiverTransfer, leavesToClaim);
      } catch (e) {
        hasError = true;
      }
      expect(hasError).toBe(true);

      // Sign an exit tx and broadcast
      const signedExitTx = await faucet.signFaucetCoin(
        exitTx,
        faucetCoin.txout,
        faucetCoin.key
      );

      await faucet.broadcastTx(signedExitTx.hex);

      // Make sure the exit tx gets enough confirmations
      const randomKey = secp256k1.utils.randomPrivateKey();
      const randomPubKey = secp256k1.getPublicKey(randomKey);
      const randomAddress = getP2TRAddressFromPublicKey(
        randomPubKey,
        Network.REGTEST
      );
      // Confirm extra buffer to scan more blocks than needed
      // So that we don't race the chain watcher in this test
      await faucet.generateToAddress(30, randomAddress);

      // Claim leaf
      await sspWallet._claimTransfer(receiverTransfer, leavesToClaim);
    },
    30000
  );
});
